### ISSUE TITLE
thread: moving toward granular critical section and multiple VCI

### ISSUE THREAD

**Issue by hzhou** [pmodels/mpich#4096]
To meet the demand of MPI + X, we need move from the current global critical section model into more granular ones and allow multiple VCI to proceed simultaneously. This issue is to track the design and progress.

I would like to point out that both large critical section and granular critical section have their merits. With large critical section, the merit is its simplicity. To be able to constantly evolve the library to meet expected performance and to add new features, we need be able to reasonably comprehend the code. Keeping a simple design is important in the long run. However, as we break the large critical section from underneath, for example, when we added yield for progress, the simplicity is abruptly lost.

With granular critical sections, the merit is it promotes parallelism by default, crucial for performance. However, it naturally makes code more complex and conflict can easily arise. For example, there is no easy way to yield, and frequent locking and unlocking adds up overhead.

Therefore, I think a good strategy is something in between. First, we recognize performance-critical path and non-performance-critical path. The performance-critical path is the progress path, including posting to and checking on the progress. This path need use the smallest granularity to ensure as much communication can happen in parallel as the hardware can accommodate. For this to work, the progress code need be simple and direct, and taking up locks only where is needed, even at device layer. It is important to prevent higher layer progress blocking each other.

For non-performance critical path, such as init/finalize, datatype commit/free, communicator creation/free, I think large per-component critical section makes sense. The critical section will be entered at upper MPIR-layer and all the lower implementation can assume the code is serialized relative to per-component. Since they take different component locks, different components will not block each other. The communications used inside the component won't block each other either can can be freely used. Only the functions that is required by MPI-standard to be strictly serialized need use the `ALL_FUNC_MUTEX`, that is init/finalize (anything else?), all other will use their per component locks.

The above described is not exactly the `per-object` model, nor exactly the `per-vci` model, but rather a hybrid of the two. Feel free to discuss or criticize.

**Comment by hzhou** [comment]
Tagging @pavanbalaji @rzambre @sagarth @raffenet @hajimefu for discussions.

**Comment by hzhou** [comment]
Superseded by issue #4219
